modules/core/src/smithy4s/http/UrlForm.scala

```diff
@@ -0,0 +1,146 @@
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s
+package http
+
+import java.net.URLEncoder
+import java.nio.charset.StandardCharsets
+
+import smithy4s.codecs.PayloadPath
+import smithy4s.codecs.PayloadPath.Segment
+import scala.collection.mutable
+import smithy4s.schema.CachedSchemaCompiler
+import smithy4s.http.internals.UrlFormDataEncoder
+import smithy4s.http.internals.UrlFormDataEncoderSchemaVisitor
+
+import smithy4s.schema.Schema
+
+private[smithy4s] final case class UrlForm(
+    formData: UrlForm.FormData.MultipleValues
+)
+
+private[smithy4s] object UrlForm {
+
+  sealed trait FormData extends Product with Serializable {
+    def prepend(segment: PayloadPath.Segment): FormData
+    def render: String
+    def toPathedValues: Vector[FormData.PathedValue]
+    def widen: FormData = this
+  }
+  object FormData {
+    case object Empty extends FormData {
+
+      override def prepend(segment: PayloadPath.Segment): FormData = this
+
+      override def render: String = ""
+
+      override def toPathedValues: Vector[FormData.PathedValue] = Vector.empty
+
+    }
+
+    // TODO: Rename as Value, replace uses by PathedValue?
+    final case class SimpleValue(string: String) extends FormData {
+
+      override def prepend(segment: PayloadPath.Segment): PathedValue =
+        PathedValue(PayloadPath(segment), maybeValue = Some(string))
+
+      override def render: String =
+        URLEncoder.encode(string, StandardCharsets.UTF_8.name())
+
+      override def toPathedValues: Vector[FormData.PathedValue] = Vector.empty
+
+    }
+
+    object PathedValue {
+      def apply(path: PayloadPath, value: String): PathedValue =
+        PathedValue(path, maybeValue = Some(value))
+    }
+
+    final case class PathedValue(path: PayloadPath, maybeValue: Option[String])
+        extends FormData {
+
+      override def prepend(segment: PayloadPath.Segment): PathedValue =
+        copy(path.prepend(segment), maybeValue)
+
+      override def render: String = {
+        val lastIndex = path.segments.size - 1
+        val renderedKey = path.segments.zipWithIndex
+          .foldLeft(new mutable.StringBuilder) {
+
+            case (builder, (Segment.Label(label), i)) if i < lastIndex =>
+              builder.append(
+                URLEncoder.encode(label, StandardCharsets.UTF_8.name())
+              )
+              builder.append('.')
+
+            case (builder, (Segment.Index(index), i)) if i < lastIndex =>
+              builder.append(index)
+              builder.append('.')
+
+            case (builder, (Segment.Label(label), _)) =>
+              builder.append(
+                URLEncoder.encode(label, StandardCharsets.UTF_8.name())
+              )
+
+            case (builder, (Segment.Index(index), _)) =>
+              builder.append(index)
+          }
+          .toString()
```

Comment:
I'd rather this was an imperative for-loop, and avoided `zipWithIndex` to save on allocation. Moreover, mixing folds with mutable-data (StringBuilder) is a code-smell.